### DISCUSSION TITLE
Why do root and modal need matching targets?

### DISCUSSION THREAD

**Discussion by LaundroMat** [unpoly/unpoly#648] [Q&A] [answered]
I'm struggling with using forms in modals. My typical code looks like

```
<a
    href="/page-with-form"
    up-layer="new"
    up-target="#form"
    up-accept-location="{{ request.path }}" 
    up-on-accepted="up.reload('#container')"
</a>
```

/page-with-form is a full page render, and I only want to render the `<div id="form">` part, hence `up-target="#form"`. The relevant part looks like this:

```
<div id="form">
    <form up-submit method="post">
    ... # input fields and submit button
    </form>
</div>

```

I understand from the docs that the page with the link to the form and /page-with-form must *both* have an element #form (in order to prevent an up.CannotMatch error), but I don't understand why that is. I only want to render the form part of /page-with-form if it's rendered in a modal, and it's pretty obvious why the page linking to it does not have a #form element.

Update: I tried fixing this with 

```
<form up-submit up-layer="root" method="post">
```

but then my whole root page is reloaded, scrolling back to the top of the page. So I just have a `<div id="content">` on each page, and target that. Still, my question remains why the same element must be present on both pages. I have a feeling that understanding this will help me grasp this wonderful library better.

**Comment by adam12** [comment]
Are you able to make an example using the [empty Glitch project](https://glitch.com/edit/#!/empty-unpoly-project) to demonstrate your issue?

It's true that when linking to a fragment, Unpoly will need to know where in the DOM the target lives so it can be updated. This isn't always true tho, as Unpoly will construct a tree in the DOM for the modal content and happily slot your target into the modal DOM constructed.

Do you see the modal pop-up, but then have an issue with the form submission inside the modal?

**Comment by LaundroMat** [comment]
I'm pretty sure I'm doing something wrong, but [here's the glitch](https://glitch.com/edit/#!/jazzy-tricolor-wolf) that shows the issue. (At least I hope it does, because I can't find how to enable up.log).

**Comment by adam12** [comment, marked as answer]
What I believe is happening is that when you target `#form` and open the new layer, that creates a scenario with no common ancestor in the modal tree and the root tree. When Unpoly submits the form, you do not provide a target so it uses the default selection of targets it knows about, which includes `[up-main]`, `main`, etc.

The response body does include a selector that matches the document root (see my remix) but Unpoly won't look in that layer by default.

A quick solution here is to tell Unpoly that it can look outside of the current layer. You can do this by applying `up-layer="current or parent"` to your `<form>` element. You may need to apply `[up-main]` to an element you consider the primary container element or modify the default selectors for `up.fragment.config.mainTargets` to include a custom selector.

See my remix: https://glitch.com/edit/#!/little-spiffy-beef?path=index.html%3A18%3A28

Instead of targetting arbitrary selectors, you might find it easier to configure [`[up-main]`](https://unpoly.com/up-main) on specific elements and then let Unpoly figure it out for you.

**Reply by LaundroMat** [reply]
Cool, thanks for the extensive reply and keep up the great work!
